kernel/thread.c

```diff
@@ -1050,7 +1050,7 @@ void z_thread_mark_switched_out(void)
 	diff = timing_cycles_get(&thread->rt_stats.last_switched_in, &now);
 #else
 	now = k_cycle_get_32();
-	diff = (uint64_t)now - thread->rt_stats.last_switched_in;
+	diff = (uint64_t)(now - thread->rt_stats.last_switched_in);
```

Comment:
Not sure I'm understanding why this fixes a bug.  What's the input that causes the wrong behavior?

This will differ from the original on 32 bit overflow.  So if the subtraction yields a value <0 or >=2^31, the old code would have produced a 64 bit value with high bits set that would look negative when cast to a signed value.  Now it will return with all the high bits zero.  Generally the former code is what you want: do all your math in the highest precision that will be used by the output.

Here, as long as the counter is monotonic, it's hard to see why it would matter except as an optimization on 32 bit platforms.

Reply:
Clarifying: the old code would **correctly** represent deltas higher than 2^31, now it will wrap in a way that doesn't correspond to the way the variable type of "diff" normally behaves.